### DISCUSSION TITLE
Training time Conv-TasNet for Different Encoder / Decoder

### DISCUSSION THREAD

**Discussion by cemox35** [asteroid-team/asteroid#582] [Q&A] [answered]
Hello there,

When I used the "free" enc/dec, 1 epoch takes nearly 1 hour, however, if I used the "STFT" enc/dec, It takes only 10 minutes. Is that normal and why is there so much difference?

**Comment by hangtingchen** [comment, marked as answer]
The settings you used may lead to a large time resolution and a large feature space. For example, win_len=16, stride=8, nfilters=512 for 8k sampling rate means that every 1ms will generate a 512-dim feat. As for stft, win_len=512, stride=256, nfilters=512 is reasonable, which means every 32ms yielding a 512dim feat, 1/32 of free encoder.

**Reply by cemox35** [reply]
How can I find the window size as ms?

**Reply by hangtingchen** [reply]
local/conf.yml (as an example):
filterbank:
  kernel_size: 16
16 represents the samples included in a window. If your audios are 8k, the window size is 16/8k=2e-3s=2ms

**Reply by cemox35** [reply]
I got it, thank you! If kernel size(i think same with win_len) is 16 in 8k sample rate, the window size is 2ms. If win_len is 512 in 8k sample rate, I found a 64ms with 512/8k. There is one point I don't understand. I think stride affects the "every 1ms will generate 512-dim" because half of win_len. If stride will change, N-dim feat generation time can change.

**Reply by hangtingchen** [reply]
Yes, you are right. For example, you can use settings "kernel_size=512, stride=1, nfilters=512" and "kernel_size=2, stride=1, nfilters=512". These two conditions will generate nearly the same shape of the feature map, but the former one will be slower due to a larger window length.
We often use 1/2 or 1/4 kernel size as the stride. For stft, the nfilters is usually equal to kernel size according to the Nyquist theorem. For a free encoder (or called learnable encoders), it prefers a redundant dictionary (much more filters).

**Reply by cemox35** [reply]
Thank you for the information you provided. I have another question 😄  I used Conv-Tasnet that is predefined in asteroid and " n_filters = 512, kernel_size = 192, stride = 128, fb_name = "stft" and sample rate is 8k. When I trained this model, after 15 epoch its "val_loss" reach the 13.7 SI-SNR score. I also used "ReduceLROnPlateau" but it cannot improve. What should I do?

**Reply by hangtingchen** [reply]
What's your goal of SI-SNR? The Conv-TasNet can reach SISNR >15dB if you use "n_filters = 512, kernel_size = 16, stride = 8, fb_name=free "

**Reply by cemox35** [reply]
Actually, My first goal decreases the training speed then acceptable SI-SNR score. If I use "kernel size=128" the SI-SNR score is nearly 16.6 in VBD dataset. Thank you for your all information.
